packages/eslint-plugin/src/rules/no-shadow.ts

```diff
@@ -24,6 +24,9 @@ type Options = [
   },
 ];
 
+const SENTINEL_TYPE =
+  /^(?:(?:Function|Class)(?:Declaration|Expression)|ArrowFunctionExpression|CatchClause|ImportDeclaration|ExportNamedDeclaration)$/u;
```

Comment:
*personally* I hate this style of check.
This is so opaque and complicated to parse and understand.

```suggestion
const SENTINEL_TYPE: ReadonlySet<AST_NODE_TYPE> = new Set([
  AST_NODE_TYPE.ArrowFunctionExpression,
  AST_NODE_TYPE.CatchClause,
  AST_NODE_TYPE.ClassDeclaration,
  AST_NODE_TYPE.ClassExpression,
  AST_NODE_TYPE.ExportNamedDeclaration,
  AST_NODE_TYPE.FunctionDeclaration,
  AST_NODE_TYPE.FunctionExpression,
  AST_NODE_TYPE.ImportDeclaration,
]);
```